### ISSUE TITLE
SelfAssignment false positive with a cast

### ISSUE THREAD

**Issue by ksobolew** [google/error-prone#2196]
### Description of the problem / feature request:

The SelfAssignment checker complains about the following line(s):

```java
// convert from 0 to 255 to -128 to 127 by converting to a signed byte
delta = (byte) delta;
```

As you can see from the comment, this is in fact not an effect-free assignment, so I believe that this is a false positive.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Have a statement like the following:

```java
field = (byte) field;
```

where `field` is a class field of type `int` (or any integer type other than `byte`). Probably doesn't even need to be a field.

### What version of Error Prone are you using?

2.5.1

###  Have you found anything relevant by searching the web?

No.

**Comment by cushon** [comment]
I agree this isn't a bug. The check was updated to check for `x = (y) x` in 7bce78f66da18452678480aa119737c3476e398b. At the time, we looked at examples of that in our code base, and found a handful that were deliberately truncating the type and other examples that were mistakes. We found the deliberate examples were often clearer when rewritten without assignment, so we made a deliberate decision to discourage this too. But that was a more subjective style/practices decision than the cases without a cast.

**Comment by ksobolew** [comment]
I see. Thing is, I tried to come up with an alternative and nothing seems right. Do you have any suggestions what would be a better alternative? Actually, now that I think of it, `delta =- 128` might do the trick.

**Comment by ksobolew** [comment]
> `delta =- 128` might do the trick

Or not. This is a different thing.

**Comment by ksobolew** [comment]
But `delta = delta << 24 >> 24` has the same semantics.

**Comment by cushon** [comment]
> Do you have any suggestions what would be a better alternative?

The sort of alternatives we used were refactoring to extra a method and avoid the reassignment, or define another variable, e.g.:

```diff
-            delta = input.read();
+            int r = input.read();
-            if (delta == -1) {
+            if (r == -1) {            
                throw new OrcCorruptionException(input.getOrcDataSourceId(), "End of stream in RLE Integer");
            }

            // convert from 0 to 255 to -128 to 127 by converting to a signed byte
-            delta = (byte) delta;
+            delta = (byte) r;
```
